File: 00085.py
class Solution:
    def maximalRectangle(self, matrix: list[list[str]]) -> int:
        row_len = len(matrix[0])
        max_area = 0
        
        left = [0 for i in range(row_len)]
        right = [len(matrix[0]) for i in range(row_len)]
        height = [0 for i in range(row_len)]
        
        for row in matrix:
            
            # count height of position i
            for i,col in enumerate(row):
                height[i] = height[i]+1 if col == '1' else 0

            # count left boundary of height[i]
            cur_left = 0
            for i,col in enumerate(row):
                if col == '1':
                    left[i] = max(left[i],cur_left)
                else:
                    left[i] = 0
                    cur_left = i+1

            # count right boundary of height[i]
            cur_right = row_len-1
            for i,col in reversed(list(enumerate(row))):
                if col == '1':
                    right[i] = min(right[i],cur_right)
                else:
                    right[i] = row_len
                    cur_right = i-1
            
            # count max area
            for i in range(row_len):
                area = ((right[i]+1)- left[i])*height[i]
                max_area = max(max_area,area) 
            
        return max_area

if __name__ == '__main__':
    a = Solution()
    print(a.maximalRectangle([["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]))

File: ProblemList/00085.cpp
#include <vector>
#include <stack>
using namespace std;
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        vector<int> heights(n,0);
        int ans = 0;
        for(int i=0; i<m; i++){
            
            // update heights
            for(int j=0; j<n; j++){
                if(matrix[i][j] == '1') heights[j] += 1;
                else heights[j] = 0;
            }

            stack<int> st;
            vector<int> left(n,-1);
            vector<int> right(n,n);

            for(int j=n-1; j>=0; j--){
                while(!st.empty() && heights[st.top()] > heights[j]){
                    left[st.top()] = j;
                    st.pop();
                }
                st.push(j);
            }

            while(!st.empty()) st.pop();

            for(int j=0; j<n; j++){
                while(!st.empty() && heights[st.top()] > heights[j]){
                    right[st.top()] = j;
                    st.pop();
                }
                st.push(j);
            }

            for(int j=0; j<n; j++){
                ans = max(ans, (right[j]-left[j]-1)*heights[j]);
            }
        }
        return ans;
    }
};